README.md
# vault-selinux-policies

This repo contains a baseline SELinux Targeted Policy, and CircleCI scripts to package the policy into RPMs (targeting CentOS and Fedora) for [HashiCorp Vault](https://www.vaultproject.io).

It is _not_ recommended to run this in Production without extensive testing first!

-----

**Please note**: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault or these SELinux Policies, _please responsibly disclose_ by contacting us at [security@hashicorp.com](mailto:security@hashicorp.com).

-----

## Overview

This repo holds the raw SELinux policy files, the packaging scripts _and_ validation scripts to package and validate via CircleCI.

### Caveats

* The SELinux policy file contexts, used as part of SELinux labelling, map to where the packaged versions of Vault install their config, data and log files. The Vault packages are available from: https://learn.hashicorp.com/tutorials/vault/getting-started-install
* The SELinux policies do allow fairly open outbound network traffic, although not outbound HTTP access by default.
* The SELinux policies haven't been tested with Vault+Consul, or other storage engines, except for Integrated Storage.
* The current policy is in Enforced Mode (which will block and audit), but can be edited into Permissive Mode (which will audit only).
* The packaged installers have only been tested with CentOS7, CentOS8, Fedora 31, Fedora 32 and Fedora 33.

## Getting Started

If you've installed Vault using our [Linux Packages](https://learn.hashicorp.com/tutorials/vault/getting-started-install) then the quickest way to install the Targeted SELinux Policy is to download the RPM for your distribution from https://github.com/hashicorp/vault-selinux-policies/releases and install that.

If you have a customized Vault installation, then you will need to clone this repo and read on.

### Booleans

While the current baseline provides fairly open access, there are some features that are gated by SELinux [Booleans](https://wiki.gentoo.org/wiki/SELinux/Tutorials/Using_SELinux_booleans).

* `vault_outbound_udp_dns` - if set will allow Vault to query DNS via UDP
* `vault_outbound_http` - if set will allow Vault to send outbound HTTP requests

To enable the booleans:

```sh
sudo setsebool vault_outbound_udp_dns on
sudo setsebool vault_outbound_http on
```

If you'd like to persist these setting:

```sh
sudo setsebool -P vault_outbound_udp_dns on
sudo setsebool -P vault_outbound_http on
```

## Editing the source

### Layout

The [products/vault_selinux/](products/vault_selinux/) folder contains the raw SELinux config files, plus [package.sh](products/vault_selinux/package.sh). This script gets executed by [circle](.circleci/config.yml).

The [products/vault_selinux/](products/vault_selinux/) folder contains the [ci/validate.sh](products/vault_selinux/ci/validate.sh) script. This gets executed by [circle](.circleci/config.yml) too.

To edit the underlying SELinux policies, edit the `vault*` files in [products/vault_selinux/](products/vault_selinux/).

The file context mappings are in `vault.fc` - the interfaces are in `vault.if` and the main policy is in `vault.te`.

The `vault.sh` is used to compile, install and bundle the policy into an RPM file. This is the main way to update the policy on the system, and was generated with `sepolicy generate --init -n vault /usr/sbin/vault`

The `vault_selinux.spec` is also generated by the above, and is used to create the RPM. Although this file has been modified.

## CI

Pushing to the main branch of this repo will execute CircleCI jobs: https://app.circleci.com/pipelines/github/hashicorp/vault-selinux-policies

These jobs will save RPM artifacts in the package steps, one for CentOS, and one for Fedora.

## Testing locally

This has only been tested on CentOS and Fedora, and requires some pre-requisites. The AWS steps below offer a more thorough example of how to test this on CentOS and Fedora.

First, clone this repo.

Then:

```sh
sudo make local-package
```

To re-install, after making changes to the SELinux files, you can re-run this script.

## Testing on AWS

To test that the RPM packages are installable the `terraform` directory contains logic to spin up 2 Centos instances.

```sh
cd terraform
make up
```

SSH to the instances then run:

```sh
sudo cloud-init status --wait
```

Wait until the cloud-init stuff has finished, this will install Vault from the latest available package.

SCP the appropriate RPM to the instances.

Then from the instances (or dnf/yum install), for example:

```sh
sudo rpm -ivh vault_selinux-1.1-1.el7.noarch.rpm
```

You can check that the policy has applied properly by checking appropriate `vault_t` contexts in:

```sh
ls -alZ /opt/vault
ls -alZ /var/log/vault
ls -alZ /usr/sbin/vault
ls -alZ /etc/vault.d
```

Or by running `semanage` commands, similar to our [validation](products/vault_selinux/ci/validate.sh) logic:

```sh
sudo semanage module -l | grep vault
sudo semanage port -l | grep vault_cluster_port_t
```

You should then be able to start the vault server on the instance:

```sh
sudo systemctl start vault.service
```

From the centos user you can:
```sh
export VAULT_ADDR=http://127.0.0.1:8200
vault status
```

Hopefully, if you tail the audit log you shouldn't see AVC errors for vault:

```sh
sudo tail -f /var/log/audit/audit.log | grep vault
```

After you've initialized and unsealed vault you can also enable the audit engine to write to /var/log/vault/vault.log

```sh
vault audit enable file file_path=/var/log/vault/vault.log
```

Don't forget to tear down your infra after:

```sh
make down
```

from the terraform folder.

## End to End Testing on AWS

First, ensure you can `make up` and `make down` as above.

Second, place a copy of the CentOS RPM into the `terraform` folder as `vault_selinux.rpm`.

Then run `make integration` from within the `terraform` folder. This will spin up the cluster, as above, then interact with the instances, deploying the RPM, and ensuring that they can function.


CHANGELOG.md
## 0.1.5
### Jan 6th, 2021

CHANGES:

* Simplified local testing with the `local-package` Makefile target

## 0.1.4
### Nov 17th, 2020

CHANGES:

* Packaging for CentOS7, CentOS8, Fedora31, Fedora32, and Fedora33

## 0.1.3
### Nov 12th, 2020

CHANGES:

* Added booleans for outbound udp/dns and http

## 0.1.2
### Sep 2nd, 2020

CHANGES:

* Updated file contexts to align with file locations from HashiCorp Vault RPM package installers

## 0.1.1
### Aug 28th, 2020

CHANGES:

* Added outbound comms permissions to allow for node-to-node comms

## 0.1.0
### Aug 12th, 2020

CHANGES:

* First iteration


terraform/e2e.sh
#!/bin/bash

set -e

INSTANCE_IP=""
INSTANCE2_IP=""
ROOT_TOKEN=""

SCRIPT_NAME="e2e.sh"

function log {
  local -r level="$1"
  local -r message="$2"
  local -r timestamp=$(date +"%Y-%m-%d %H:%M:%S")
  >&2 echo -e "${timestamp} [${level}] [$SCRIPT_NAME] ${message}"
}

function log_info {
  local -r message="$1"
  log "INFO" "$message"
}

function log_warn {
  local -r message="$1"
  log "WARN" "$message"
}

function log_error {
  local -r message="$1"
  log "ERROR" "$message"
}

function assert_not_empty {
  local -r arg_name="$1"
  local -r arg_value="$2"

  if [[ -z "$arg_value" ]]; then
    log_error "The value for '$arg_name' cannot be empty"
    print_usage
    exit 1
  fi
}

# A retry function that attempts to run a command a number of times and returns the output
function retry {
  local -r cmd="$1"
  local -r description="$2"

  for i in $(seq 1 15); do
    log_info "$description"

    # The boolean operations with the exit status are there to temporarily circumvent the "set -e" at the
    # beginning of this script which exits the script immediatelly for error status while not losing the exit status code
    output=$(eval "$cmd") && exit_status=0 || exit_status=$?
    log_info "$output"
    if [[ $exit_status -eq 0 ]]; then
      echo "$output"
      return
    fi
    log_warn "$description failed. Will sleep for 10 seconds and try again."
    sleep 10
  done;

  log_error "$description failed after 15 attempts."
  exit $exit_status
}

function get_instance_ip() {
  local -r tfitem="$1"

  if [[ -z "$tfitem" ]]; then
    log_error "Need to provide a TF output value.."
    exit 1
  fi

  tf_out=$(terraform output ${tfitem})
  if [[ $tf_out == "" ]]; then
    log_error "Can't find instance ip from TF, quitting..."
    exit 1
  fi

  log_info "Found ${tfitem} ${tf_out}"

  if [[ "$tfitem" == "instance_ip" ]]; then
    INSTANCE_IP=$tf_out
  fi

  if [[ "$tfitem" == "instance2_ip" ]]; then
    INSTANCE2_IP=$tf_out
  fi
}

function wait_for_instance() {
 local -r ip="$1"

 if [[ -z "$ip" ]]; then
   log_error "Need to provide an IP to wait for instance"
   exit 1
 fi

  retry "ssh -o 'StrictHostKeyChecking=no' centos@${ip} sudo cloud-init status --wait" "Waiting for cloud-init..."
}

function start_vault() {
  local -r ip="$1"
  local init="$2"
  local seal="$3"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to start_vault"
    exit 1
  fi

  if [[ -z "$init" ]]; then
    log_info "No init target state specified, checking for false"
    init="false"
  fi

  if [[ -z "$seal" ]]; then
    log_info "No seal target state specified, checking for true"
    seal="true"
  fi


  log_info "Starting Vault on ${ip}"
  ssh centos@${ip} sudo systemctl start vault
  log_info "Sleeping for 10, waiting for Vault to start..."
  sleep 10

  status=$(ssh centos@"$ip" VAULT_ADDR=http://127.0.0.1:8200 vault status -format=json | jq -r '.')

  actual_init=$(echo $status | jq -r '.initialized')
  actual_seal=$(echo $status | jq -r '.sealed')

  if [[ "$init" != "$actual_init" ]]; then
    log_error "Init status invalid: ${init} did not equal ${actual_init}"
    exit 1
  fi

  if [[ "$seal" != "$actual_seal" ]]; then
    log_error "Seal status: ${seal} did not equal ${actual_seal}"
    exit 1
  fi

  log_info "Vault appears to have started.."

}

function stop_vault() {
  local -r ip="$1"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to stop_vault"
    exit 1
  fi

  log_info "Stopping vault service on ${ip}"

  ssh centos@${ip} sudo systemctl stop vault
}

function vault_init() {
  local -r ip="$1"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to init vault"
    exit 1
  fi

  log_info "Initializing Vault on ${ip}"

  init=$(ssh centos@${ip} VAULT_ADDR=http://127.0.0.1:8200 vault operator init -format=json | jq -r '.')
  ROOT_TOKEN=$(echo $init | jq -r '.root_token')

  log_info "Got ROOT Token: ${ROOT_TOKEN}"
}

function vault_list_peers() {
  local -r ip="$1"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to check peers"
    exit 1
  fi

  log_info "Fetching raft peer list"

  peers=$(ssh centos@${ip} VAULT_ADDR=http://127.0.0.1:8200 VAULT_TOKEN=${ROOT_TOKEN} vault operator raft list-peers -format=json | jq -r '.')
  peer_count=$(echo $peers | jq -r '.data.config.servers | length')

  if [[ "$peer_count" != "2" ]]; then
    log_error "Peer count doesn't equal 2, instead it equals ${peer_count}"
    exit 1
  fi

  log_info "Looks like we have two peers, which is correct!!"
}

function push_file() {
  local -r ip="$1"
  local -r file="$2"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to push file to"
    exit 1
  fi

  if [[ -z "$file" ]]; then
    log_error "Need to provide path to file"
    exit 1
  fi

  log_info "Pushing ${file} to ${ip}:selinux.rpm"

  scp ${file} centos@${ip}:selinux.rpm
}

function install_rpm() {
  local -r ip="$1"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to install the uploaded RPM"
    exit 1
  fi

  log_info "Installing the ~/selinux.rpm on ${ip}"

  ssh centos@${ip} sudo yum install -y ./selinux.rpm
}

function update_selinux_outbound() {
  local -r ip="$1"
  local -r state="$2"

  if [[ -z "$ip" ]]; then
    log_error "Need to provide IP to update selinux bools"
    exit 1
  fi

  if [[ -z "$state" ]]; then
    log_error "Need to provide target state of on or off"
    exit 1
  fi

  log_info "Updating selinux policies to allow outbound DNS and HTTP"

  ssh centos@${ip} sudo setsebool vault_outbound_udp_dns ${state}
  ssh centos@${ip} sudo setsebool vault_outbound_http ${state}
}

function print_usage() {
  echo
  echo "Usage: ./e2e.sh [OPTIONS]"
  echo
  echo "Options:"
  echo
  echo -e "  --rpm-file\t\tThe full path to the SELinux RPM file for testing. This is required"
  echo -e "  --help\t\tPrint help"
}

function e2e {
  local rpm_file=""

  while [[ $# > 0 ]]; do
    local key="$1"

    case "$key" in
      --rpm-file)
        rpm_file="$2"
        shift
        ;;
      --help)
        print_usage
        exit
        ;;
      *)
        log_error "Unrecognized argument: $key"
        print_usage
        exit 1
        ;;
    esac

    shift
  done

  assert_not_empty "--rpm-file" "$rpm_file"

  log_info "====================== Starting cluster"

  get_instance_ip "instance_ip"
  wait_for_instance $INSTANCE_IP

  start_vault $INSTANCE_IP
  vault_init $INSTANCE_IP

  get_instance_ip "instance2_ip"
  wait_for_instance $INSTANCE2_IP

  start_vault $INSTANCE2_IP "true" "false"

  vault_list_peers $INSTANCE_IP

  log_info "====================== Stopping cluster"

  stop_vault $INSTANCE2_IP
  stop_vault $INSTANCE_IP

  log_info "====================== Pushing RPMs to cluster"

  push_file $INSTANCE_IP $rpm_file
  install_rpm $INSTANCE_IP
  update_selinux_outbound $INSTANCE_IP "on"

  push_file $INSTANCE2_IP $rpm_file
  install_rpm $INSTANCE2_IP
  update_selinux_outbound $INSTANCE2_IP "on"

  log_info "====================== Restarting vault cluster"

  start_vault $INSTANCE_IP "true" "false"
  start_vault $INSTANCE2_IP "true" "false"

  vault_list_peers $INSTANCE_IP

  log_info "====================== End to End Testing Completed Successfully!"
}

e2e "$@"


products/vault_selinux/vault.sh
#!/bin/sh -e

DIRNAME=`dirname $0`
cd $DIRNAME
USAGE="$0 [ --update ]"
if [ `id -u` != 0 ]; then
echo 'You must be root to run this script'
exit 1
fi

if [ $# -eq 1 ]; then
	if [ "$1" = "--update" ] ; then
		time=`ls -l --time-style="+%x %X" vault.te | awk '{ printf "%s %s", $6, $7 }'`
		rules=`ausearch --start $time -m avc --raw -se vault`
		if [ x"$rules" != "x" ] ; then
			echo "Found avc's to update policy with"
			echo -e "$rules" | audit2allow -R
			echo "Do you want these changes added to policy [y/n]?"
			read ANS
			if [ "$ANS" = "y" -o "$ANS" = "Y" ] ; then
				echo "Updating policy"
				echo -e "$rules" | audit2allow -R >> vault.te
				# Fall though and rebuild policy
			else
				exit 0
			fi
		else
			echo "No new avcs found"
			exit 0
		fi
	else
		echo -e $USAGE
		exit 1
	fi
elif [ $# -ge 2 ] ; then
	echo -e $USAGE
	exit 1
fi

echo "Building and Loading Policy"
set -x
make -f /usr/share/selinux/devel/Makefile vault.pp || exit
/usr/sbin/semodule -i vault.pp

# Generate a man page off the installed module
sepolicy manpage -p . -d vault_t
# Fixing the file context on /usr/sbin/vault
/sbin/restorecon -F -R -v -i /usr/bin/vault
# Fixing the file context on /var/log/vault
/sbin/restorecon -F -R -v -i /var/log/vault
# Fixing the file context on /opt/vault
/sbin/restorecon -F -R -v -i /opt/vault
# Fixing the file context on /etc/vault.d
/sbin/restorecon -F -R -v -i /etc/vault.d
# Labeling 8201
/sbin/semanage port -a -t vault_cluster_port_t -p tcp 8201
# Generate a rpm package for the newly generated policy

pwd=$(pwd)
rpmbuild --define "_sourcedir ${pwd}" --define "_specdir ${pwd}" --define "_builddir ${pwd}" --define "_srcrpmdir ${pwd}" --define "_rpmdir ${pwd}" --define "_buildrootdir ${pwd}/.build"  -ba vault_selinux.spec


products/vault_selinux/package.sh
#!/usr/bin/env bash
set -xeu pipefail

HC_VERSION=${HC_VERSION}
PACKAGE_ITERATION=${HC_PACKAGE_ITERATION:-1}
LOCAL=${LOCAL_PACKAGE:-0}

PRODUCT_NAME="vault_selinux"

OUTPUT_PATH=$(pwd)

if [[ "$LOCAL" == "1" ]]; then
  echo "Performing a local package install"
  HC_VERSION="0.0.1"
else
  # Create temporary workspace
  echo "Performing CI package install"
  echo "Creating temporary workspace"
  mkdir pkg_tmp

  cp ./vault.fc ./pkg_tmp/vault.fc
  cp ./vault.if ./pkg_tmp/vault.if
  cp ./vault.sh ./pkg_tmp/vault.sh
  cp ./vault.te ./pkg_tmp/vault.te
  cp ./vault_selinux.spec ./pkg_tmp/vault_selinux.spec

  PACKAGE_DIR=$(cd pkg_tmp; pwd)
  cd $PACKAGE_DIR
fi

# @TODO: I'm sure there are better ways to build RPM packages for Fedora & Centos
# Currently I'm doing this in two different containers, with some %if logic in the
# vault_selinux.spec file

# The current package CircleCI config has a matrix for a Centos & Fedora container

# Check for CentOS or Fedora

if [ -f /etc/os-release ]; then
  . /etc/os-release
  OS=$NAME
fi

if [[ $OS == *"CentOS"* ]]; then
  echo "Detected CentOS"

  # Install Vault RPM
  echo "yum installing Vault"
  yum install -y yum-utils
  yum-config-manager --add-repo https://rpm.releases.hashicorp.com/RHEL/hashicorp.repo
  yum -y install vault

  # Install other deps
  echo "yum installing other stuff"
  yum -y install policycoreutils-devel setools-console rpm-build selinux-policy-devel selinux-policy-targeted

elif [[ $OS == *"Fedora"* ]]; then
  echo "Detected Fedora"

  # Install Vault RPM
  echo "dnf installing Vault"
  dnf install -y dnf-plugins-core
  dnf config-manager --add-repo https://rpm.releases.hashicorp.com/fedora/hashicorp.repo
  dnf -y install vault

  # Install other deps
  echo "dnf installing other stuff"
  dnf -y install policycoreutils-devel setools-console rpm-build

fi

echo "Updating #VERSION# in vault.te and vault_selinux.spec"
sed -i "s^#VERSION#^${HC_VERSION}^g" vault.te
sed -i "s^#VERSION#^${HC_VERSION}^g" vault_selinux.spec

# Run the sepolicy builder
if [[ "$LOCAL" == "1" ]]; then
  sudo sh ./vault.sh
else
  sh ./vault.sh
  cp *.rpm $OUTPUT_PATH
  cp noarch/*.rpm $OUTPUT_PATH

  # Cleanup
  cd $OUTPUT_PATH
  rm -rf $PACKAGE_DIR
fi


products/vault_selinux/ci/validate.sh
#!/usr/bin/env bash
set -eux

# This script is executed from a circleci machine executor to
# validate the Linux rpm packages for vault_selinux

WORKDIR=$(pwd)

# Let our docker containers write to mounted volumes
if ! [ -z "${CI:-""}" ]; then
  chmod a+rwx -R ./
fi

if [[ $TARGET_IMAGE == $IMAGE_CENTOS8_SYSTEM ]]; then

  # Run CentOS8 rpm validation
  echo "Testing on Centos8"
  CENTOS_ID=$(docker run -d -v $WORKDIR:/app -e HC_PRODUCT=$HC_PRODUCT -e HC_VERSION=$HC_VERSION \
    --entrypoint="" -w="/app" --privileged $IMAGE_CENTOS8_SYSTEM /usr/sbin/init)
  # Wait for CentOS to spin up
  sleep 1
  docker exec $CENTOS_ID yum install -y libselinux-utils policycoreutils policycoreutils-python-utils selinux-policy-targeted
  docker exec $CENTOS_ID yum install -y /app/$(ls products/*/*el8.noarch.rpm)

  docker exec $CENTOS_ID bash -c 'semanage module -l | grep vault'
  docker exec $CENTOS_ID bash -c 'semanage port -l | grep vault_cluster_port_t'
  docker exec $CENTOS_ID bash -c 'semanage boolean -l | grep vault_outbound'
  docker exec $CENTOS_ID yum remove -y vault_selinux

  docker stop $CENTOS_ID

fi

if [[ $TARGET_IMAGE == $IMAGE_CENTOS7_SYSTEM ]]; then
  # Run CentOS7 rpm validation
  echo "Testing on Centos7"
  CENTOS_ID=$(docker run -d -v $WORKDIR:/app -e HC_PRODUCT=$HC_PRODUCT -e HC_VERSION=$HC_VERSION \
    --entrypoint="" -w="/app" --privileged $IMAGE_CENTOS7_SYSTEM /usr/sbin/init)
  # Wait for CentOS to spin up
  sleep 1
  docker exec $CENTOS_ID yum install -y libselinux-utils policycoreutils policycoreutils-python selinux-policy-targeted
  docker exec $CENTOS_ID yum install -y /app/$(ls products/*/*el7.noarch.rpm)

  docker exec $CENTOS_ID bash -c 'semanage module -l | grep vault'
  docker exec $CENTOS_ID bash -c 'semanage port -l | grep vault_cluster_port_t'
  docker exec $CENTOS_ID bash -c 'semanage boolean -l | grep vault_outbound'
  docker exec $CENTOS_ID yum remove -y vault_selinux

  docker stop $CENTOS_ID

fi

if [[ $TARGET_IMAGE == $IMAGE_F31_SYSTEM ]]; then
  # Run Fedora31 rpm validation
  echo "Testing on Fedora31"
  FEDORA_ID=$(docker run -d -v $WORKDIR:/app -e HC_PRODUCT=$HC_PRODUCT -e HC_VERSION=$HC_VERSION \
    --entrypoint="" -w="/app" --privileged -ti $IMAGE_F31_SYSTEM /bin/bash)
  # Wait for Fedora to spin up
  sleep 1
  docker exec $FEDORA_ID dnf install -y libselinux-utils policycoreutils policycoreutils-python-utils selinux-policy-targeted
  docker exec $FEDORA_ID dnf install -y /app/$(ls products/*/*fc31.noarch.rpm)

  docker exec $FEDORA_ID bash -c 'semanage module -l | grep vault'
  docker exec $FEDORA_ID bash -c 'semanage port -l | grep vault_cluster_port_t'
  docker exec $FEDORA_ID bash -c 'semanage boolean -l | grep vault_outbound'
  docker exec $FEDORA_ID dnf remove -y vault_selinux

  docker stop $FEDORA_ID

fi

if [[ $TARGET_IMAGE == $IMAGE_F32_SYSTEM ]]; then
  # Run Fedora32 rpm validation
  echo "Testing on Fedora32"
  FEDORA_ID=$(docker run -d -v $WORKDIR:/app -e HC_PRODUCT=$HC_PRODUCT -e HC_VERSION=$HC_VERSION \
    --entrypoint="" -w="/app" --privileged -ti $IMAGE_F32_SYSTEM /bin/bash)
  # Wait for Fedora to spin up
  sleep 1
  docker exec $FEDORA_ID dnf install -y libselinux-utils policycoreutils policycoreutils-python-utils selinux-policy-targeted
  docker exec $FEDORA_ID dnf install -y /app/$(ls products/*/*fc32.noarch.rpm)

  docker exec $FEDORA_ID bash -c 'semanage module -l | grep vault'
  docker exec $FEDORA_ID bash -c 'semanage port -l | grep vault_cluster_port_t'
  docker exec $FEDORA_ID bash -c 'semanage boolean -l | grep vault_outbound'
  docker exec $FEDORA_ID dnf remove -y vault_selinux

  docker stop $FEDORA_ID

fi

if [[ $TARGET_IMAGE == $IMAGE_F33_SYSTEM ]]; then
  # Run Fedora33 rpm validation
  echo "Testing on Fedora33"
  FEDORA_ID=$(docker run -d -v $WORKDIR:/app -e HC_PRODUCT=$HC_PRODUCT -e HC_VERSION=$HC_VERSION \
    --entrypoint="" -w="/app" --privileged -ti $IMAGE_F33_SYSTEM /bin/bash)
  # Wait for Fedora to spin up
  sleep 1
  docker exec $FEDORA_ID dnf install -y libselinux-utils policycoreutils policycoreutils-python-utils selinux-policy-targeted
  docker exec $FEDORA_ID dnf install -y /app/$(ls products/*/*fc33.noarch.rpm)

  docker exec $FEDORA_ID bash -c 'semanage module -l | grep vault'
  docker exec $FEDORA_ID bash -c 'semanage port -l | grep vault_cluster_port_t'
  docker exec $FEDORA_ID bash -c 'semanage boolean -l | grep vault_outbound'
  docker exec $FEDORA_ID dnf remove -y vault_selinux

  docker stop $FEDORA_ID
fi

echo "Validation tests complete."


terraform/user-data-vault.sh
#!/bin/bash
set -e

# Send the log output from this script to user-data.log, syslog, and the console
# From: https://alestic.com/2010/12/ec2-user-data-output/
exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1

sudo yum install -y yum-utils
sudo yum-config-manager --add-repo https://rpm.releases.hashicorp.com/RHEL/hashicorp.repo
sudo yum -y install vault-1.6.0-1

sudo yum -y install wget unzip tmux policycoreutils-devel setools-console rpm-build vim
export instance_id="$(curl -s http://169.254.169.254/latest/meta-data/instance-id)"
export internal_ip="$(curl -s http://169.254.169.254/latest/meta-data/local-ipv4)"
export external_ip="$(curl -s http://169.254.169.254/latest/meta-data/public-ipv4)"


sudo tee /etc/vault.d/vault.hcl > /dev/null <<EOT
ui = true

#mlock = true
#disable_mlock = true

#storage "file" {
#  path = "/opt/vault/data"
#}

storage "raft" {
  path = "/opt/vault/data"
  node_id = "$instance_id"

  retry_join {
    auto_join = "provider=aws addr_type=private_v4 tag_key=cluster_name tag_value=raft"
    auto_join_scheme = "http"
  }
}

seal "awskms" {
  region     = "${region}"
  kms_key_id = "${kms_key_id}"
}

#storage "consul" {
#  address = "127.0.0.1:8500"
#  path    = "vault"
#}

# HTTP listener
#listener "tcp" {
#  address = "127.0.0.1:8200"
#  tls_disable = 1
#}

# HTTPS listener
listener "tcp" {
  address       = "0.0.0.0:8200"
  tls_disable = 1

#  tls_cert_file = "/opt/vault/tls/tls.crt"
#  tls_key_file  = "/opt/vault/tls/tls.key"
}
api_addr = "http://0.0.0.0:8200"
cluster_addr = "http://$internal_ip:8201"
EOT

